metrics/engine/ingester.go

```diff
@@ -92,6 +97,74 @@ func (oi *outputIngester) flushMetrics() {
 				oi.metricsEngine.markObserved(sm.Metric)
 				sm.Metric.Sink.Add(sample)
 			}
+
+			oi.cardinality.Add(sample.TimeSeries)
 		}
 	}
+
+	if hit, limit := oi.cardinality.LimitHit(); hit {
+		// TODO: suggest using the Metadata API as an alternative, once it's
+		// available (e.g. move high-cardinality tags as Metadata)
+		// https://github.com/grafana/k6/issues/2766
+
+		oi.logger.Warnf("The test has generated %d time series, which is higher than the suggested limit (%d), "+
+			"and could cause high memory usage. "+
+			"Consider not using high-cardinality values like IDs as tags "+
+			"or if you need them in the URL then use the URL-grouping feature. "+
+			"See https://k6.io/docs/using-k6/tags-and-groups for details.", oi.cardinality.Count(), limit)
+	}
+}
+
+type cardinalityControl struct {
+	seen            map[metrics.TimeSeries]struct{}
+	timeSeriesLimit int
+	addsCount       int
+}
+
+func newCardinalityControl() *cardinalityControl {
+	return &cardinalityControl{
+		timeSeriesLimit: timeSeriesFirstLimit,
+		addsCount:       0,
+		seen:            make(map[metrics.TimeSeries]struct{}),
+	}
+}
+
+// Add adds the passed time series to the list of seen items.
+func (cc *cardinalityControl) Add(ts metrics.TimeSeries) {
+	cc.addsCount++
+	cc.seen[ts] = struct{}{}
```

Comment:
hmm, I know Go has some optimizations for maps that have `struct{}` values, but we need to double-check that this is not doing any allocations on the heap for every call to `Add()`, or something weird like that :thinking: 

I hope this empty struct initialization and map assignment is optimized away by the compiler, but if it's not, then something like this might end up more efficient:
```suggestion
	if _, ok := cc.seen[ts]; !ok {
		cc.seen[ts] = struct{}{}
	}
```

Reply:
Sorry for unresolving this, but my point was that I am not sure which is better, it depends on how Go optimizes `map[whatever]struct{}` operations. If it does optimize them properly, it's likely that your original solution is better.

Reply:
I went in a bit of a rabbit hole, but this can probably only be settled by a realistic benchmark, and it doesn't really matter too much either way :sweat_smile: 

FWIW, if https://godbolt.org/ is accurate, then I don't think there are any special optimizations for `struct{}` map values on the map side specifically, though there are optimizations for allocating memory. I was playing with this code, which should be fairly representative, I think:
```go
package main

import "fmt"

var Data map[*int]struct{} = make(map[*int]struct{})
//var Data map[*int]any = make(map[*int]any)
//var Data map[*int]bool = make(map[*int]bool)

func set(p *int) {
    //if _, ok := Data[p]; !ok {
        Data[p] = struct{}{}
        //Data[p] = nil
        //Data[p] = true
    //}
}

func main() {
	data := [...]int{1, 2, 3, 4, 5}
	set(&data[0])
	set(&data[1])
	set(&data[0])
	fmt.Printf("%#q", Data)
}
```

With the `if` commented out, Go will only call [`runtime.mapassign_fast64ptr()`](https://github.com/golang/go/blob/f57f02fcd519a86683c47a444e0e24a086fea8e0/src/runtime/map_fast64.go#L183-L271). If we uncomment the `if` check, it first calls [`mapaccess2_fast64()`](https://github.com/golang/go/blob/f57f02fcd519a86683c47a444e0e24a086fea8e0/src/runtime/map_fast64.go#L53-L91) and then `runtime.mapassign_fast64ptr()`, as we'd expect if there are no special optimizations. And this holds true regardless of whether we use `struct{}`, `interface{}` / `any`, or `bool`. The `any` code has extra code with `runtime.writeBarrier()`, which I didn't get to, but the only difference between `struct{}` and `bool` seems to be a `TESTB` vs `MOVB` instruction at the end.

Looking at the code of these Go runtime functions, my gut feeling is that your original solution of just "assigning" the `struct{}{}` value without the check would be somewhat faster in most cases, though I haven't actually tested it.